WPThemeReview/Sniffs/ThouShallNotUse/ReservedTemplateNameSniff.php

```diff
@@ -0,0 +1,137 @@
+<?php
+/**
+ * WPThemeReview Coding Standard.
+ *
+ * @package WPTRT\WPThemeReview
+ * @link    https://github.com/WPTRT/WPThemeReview
+ * @license https://opensource.org/licenses/MIT MIT
+ */
+
+namespace WPThemeReview\Sniffs\ThouShallNotUse;
+
+use PHP_CodeSniffer\Sniffs\Sniff;
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Util\Tokens;
+
+/**
+ * Check if the template file is using reserved file name as a part of their name.
+ *
+ * @since 0.2.0
+ */
+class ReservedTemplateNameSniff implements Sniff {
+
+	/**
+	 * Error message template.
+	 *
+	 * @var string
+	 */
+	const ERROR_MSG = 'File template should not use a reserved name as a prefix. Found: "%s".';
+
+	/**
+	 * Regex to catch the template creation comment
+	 *
+	 * @var string
+	 */
+	const TEMPLATE_REGEX = '`\/\* Template Name:`i';
+
+	/**
+	 * Found prefix in a file
+	 *
+	 * @var string
+	 */
+	protected $prefix;
+
+	/**
+	 * List of reserved template file names.
+	 *
+	 * @link https://developer.wordpress.org/themes/basics/template-hierarchy/
+	 * @link https://developer.wordpress.org/themes/template-files-section/partial-and-miscellaneous-template-files/#content-slug-php
+	 * @link https://wphierarchy.com/
+	 * @link https://en.wikipedia.org/wiki/Media_type#Naming
+	 *
+	 * @since 0.2.0
+	 *
+	 * @var array
+	 */
+	protected $reserved_file_name_prefixes = [ 'index-', 'comments-', 'front-page-', 'home-', 'header-', 'singular-', 'single-', 'page-', 'category-', 'tag-', 'taxonomy-', 'author-', 'date-', 'archive-', 'search-', 'attachment-', 'image-', '404-' ];
```

Comment:
If I read the theme handbook, the template hierarchy and the WP Core code for loading templates correctly, the only ones which would be problematic are:
* `author-`
* `category-`
* `tag-`
* `page-`

The problem occurs if the WP logic for including a template could confuse the file name for `type-$slug` and AFAICS, those are the only ones where that could happen.

A template name like `comments-something.php` would just be ignored by WP and have no effect, except when specifically `include`d by theme from one of the other files.

Reply:
Yes, I thought about this too, but didn't find any additional information about if people are doing such thing (like `index-secondary.php` or `comments-lists.php`).

But thought - well it doesn't hurt to catch these cases as well 🤷‍♂Maybe we could warn the users if they use something like `index-secondary.php`, `home-secondary.php` or `comments-something.php` that WordPress will ignore these. Although the same would happen with the above due to how the WP template hierarchy works, so I guess it makes no sense to split these two.

The biggest issue for me is that I don't have the concrete number or evidence if these sort of things happen at all (the index prefix for instance). I'll try to get all the themes and search through them to see if I'm getting any positives on this

Reply:
> So should we check for all the prefixes?

The question is: what is this sniff about ?
Is it about templates which will be interpreted as `$slug` specific by WP ? If so, then no, this should not be in this sniff or at the very least, it should have a different error code.
Those other templates will be ignored by WP natively, but can be included from within the theme from another template file, checking whether that is done is a whole other matter.

Reply:
I see. Well the original issue was with `$slug` specific so only 4 problematic ones are the ones you mentioned :+1: